File: Tipoo.Api/Models/GameSettings.cs
namespace Tipoo.Api.Models;

public class GameSettings
{
    public const int MinRoundTimeSeconds = 30;
    public const int MaxRoundTimeSeconds = 600;
    public const int MinNumberOfRounds = 2;
    public const int MaxNumberOfRounds = 20;
    public const int MinTipooLeadLimit = 10;
    public const int MaxExplanationTimeSeconds = 15;
    public const int MaxSkipLimit = 10;
    public const int MaxPoints = 10;
    public const int MinPauseBetweenRoundsSeconds = 15;
    public const int MaxPauseBetweenRoundsSeconds = 300;

    public int RoundTimeSeconds { get; set; } = 180;
    public int NumberOfRounds { get; set; } = 6;
    public int SkipLimit { get; set; } = 3;
    public bool SkipCostsPoints { get; set; }
    public int? TipooLeadLimit { get; set; }
    public int ExplanationTimeSeconds { get; set; } = 5;
    public List<string> Difficulties { get; set; } = new() { "Fácil", "Médio", "Difícil" };
    public List<string> BuzzerSounds { get; set; } = new() { "air-horn", "censura", "erro" };
    public bool RandomBuzzerSound { get; set; } = true;
    public bool PanicMode { get; set; }
    public int PointsPerCorrect { get; set; } = 1;
    public int PointsPerError { get; set; } = 1;
    public int PointsPerSkip { get; set; } = 1;
    public string StartingTeam { get; set; } = "aleatorio";
    public string TiebreakMode { get; set; } = "empatado";
    public int PauseBetweenRoundsSeconds { get; set; } = 30;
    public int ReanalysisSelectionTimeSeconds { get; set; } = 20;
    public int CardJudgingTimeSeconds { get; set; } = 10;

    public GameSettings Clone()
    {
        return new GameSettings
        {
            RoundTimeSeconds = RoundTimeSeconds,
            NumberOfRounds = NumberOfRounds,
            SkipLimit = SkipLimit,
            SkipCostsPoints = SkipCostsPoints,
            TipooLeadLimit = TipooLeadLimit,
            ExplanationTimeSeconds = ExplanationTimeSeconds,
            Difficulties = (Difficulties ?? new List<string>()).ToList(),
            BuzzerSounds = (BuzzerSounds ?? new List<string>()).ToList(),
            RandomBuzzerSound = RandomBuzzerSound,
            PanicMode = PanicMode,
            PointsPerCorrect = PointsPerCorrect,
            PointsPerError = PointsPerError,
            PointsPerSkip = PointsPerSkip,
            StartingTeam = StartingTeam,
            TiebreakMode = TiebreakMode,
            PauseBetweenRoundsSeconds = PauseBetweenRoundsSeconds,
            ReanalysisSelectionTimeSeconds = ReanalysisSelectionTimeSeconds,
            CardJudgingTimeSeconds = CardJudgingTimeSeconds
        };
    }
}

File: tipoo-client/src/app/models/game-settings.ts
export interface GameSettings {
  roundTimeSeconds: number;
  numberOfRounds: number;
  skipLimit: number;
  skipCostsPoints: boolean;
  tipooLeadLimit: number | null;
  explanationTimeSeconds: number;
  difficulties: string[];
  buzzerSounds: string[];
  randomBuzzerSound: boolean;
  panicMode: boolean;
  pointsPerCorrect: number;
  pointsPerError: number;
  pointsPerSkip: number;
  startingTeam: 'azul' | 'vermelho' | 'aleatorio';
  tiebreakMode: 'empatado' | 'rodada-extra';
  pauseBetweenRoundsSeconds: number;
}

export const DEFAULT_DIFFICULTIES = ['Fácil', 'Médio', 'Difícil'];
export const BUZZER_SOUNDS = ['air-horn', 'censura', 'erro'];
export const STARTING_TEAM_OPTIONS = ['azul', 'vermelho', 'aleatorio'] as const;
export const TIEBREAK_OPTIONS = ['empatado', 'rodada-extra'] as const;
export const DEFAULT_TIPOO_LEAD_LIMIT = 100;

export function createDefaultGameSettings(): GameSettings {
  return {
    roundTimeSeconds: 180,
    numberOfRounds: 6,
    skipLimit: 3,
    skipCostsPoints: false,
    tipooLeadLimit: null,
    explanationTimeSeconds: 5,
    difficulties: [...DEFAULT_DIFFICULTIES],
    buzzerSounds: [...BUZZER_SOUNDS],
    randomBuzzerSound: true,
    panicMode: false,
    pointsPerCorrect: 1,
    pointsPerError: 1,
    pointsPerSkip: 1,
    startingTeam: 'aleatorio',
    tiebreakMode: 'empatado',
    pauseBetweenRoundsSeconds: 30,
  };
}
